src/c_kzg_4844.c

```diff
@@ -2870,83 +2911,101 @@ static C_KZG_RET recover_cells_impl(
     /* Check that we have enough cells */
     assert(len_missing <= s->max_width / 2);
 
-    /* Calculate Z_r,I */
+    /* Compute Z(x) in monomial form */
     ret = zero_polynomial_via_multiplication(
-        zero_poly, &zero_poly_len, missing, len_missing, s
+        zero_poly_coeff, &zero_poly_len, missing, len_missing, s
     );
     if (ret != C_KZG_OK) goto out;
 
-    /* Evaluate the zero poly */
-    ret = fft_fr(zero_eval, zero_poly, s->max_width, s);
+    /* Convert Z(x) to evaluation form */
+    ret = fft_fr(zero_poly_eval, zero_poly_coeff, s->max_width, s);
     if (ret != C_KZG_OK) goto out;
 
-    /* Construct E * Z_r,I: the loop makes the evaluation polynomial */
+    /* Compute (E*Z)(x) = E(x) * Z(x) in evaluation form over the FFT domain */
     for (size_t i = 0; i < s->max_width; i++) {
         if (fr_is_null(&cells_brp[i])) {
-            poly_evaluations_with_zero[i] = FR_ZERO;
+            extended_evaluation_times_zero[i] = FR_ZERO;
         } else {
             blst_fr_mul(
-                &poly_evaluations_with_zero[i], &cells_brp[i], &zero_eval[i]
+                &extended_evaluation_times_zero[i],
+                &cells_brp[i],
+                &zero_poly_eval[i]
             );
         }
     }
 
-    /* Now inverse FFT so that poly_with_zero is (D * Z_r,I)(x) */
-    ret = ifft_fr(poly_with_zero, poly_evaluations_with_zero, s->max_width, s);
+    /* Convert (E*Z)(x) to monomial form  */
+    ret = ifft_fr(
+        extended_evaluation_times_zero_coeffs,
+        extended_evaluation_times_zero,
+        s->max_width,
+        s
+    );
     if (ret != C_KZG_OK) goto out;
 
-    /* Scale both polynomials */
-    scale_poly(poly_with_zero, s->max_width);
-    scale_poly(zero_poly, zero_poly_len);
-
     /*
      * Polynomial division by convolution: Q3 = Q1 / Q2 where
      *   Q1 = (D * Z_r,I)(k * x)
      *   Q2 = Z_r,I(k * x)
      *   Q3 = D(k * x)
      */
-    ret = fft_fr(eval_poly_with_zero, poly_with_zero, s->max_width, s);
+    ret = coset_fft_fr(
+        extended_evaluations_over_coset,
+        extended_evaluation_times_zero_coeffs,
+        s->max_width,
+        s
+    );
     if (ret != C_KZG_OK) goto out;
 
     /* We use max_width here (not zero_poly_len) intentionally */
-    ret = fft_fr(eval_zero_poly, zero_poly, s->max_width, s);
+    ret = coset_fft_fr(zero_poly_over_coset, zero_poly_coeff, s->max_width, s);
     if (ret != C_KZG_OK) goto out;
 
     /* The result of the division is Q3 */
     for (size_t i = 0; i < s->max_width; i++) {
         fr_div(
-            &eval_poly_with_zero[i], &eval_poly_with_zero[i], &eval_zero_poly[i]
+            &extended_evaluations_over_coset[i],
+            &extended_evaluations_over_coset[i],
+            &zero_poly_over_coset[i]
         );
     }
 
+    /* reconstructed_poly_over_coset is now extended_evaluations_over_coset */
```

Comment:
The `reconstructed_poly_over_coset` variable doesn't exist. Can we delete this comment?

Reply:
This was an ambiguous comment. `reconstructed_poly_over_coset` is a variable in the spec. Will reword.

Reply:
Addressed in 0edcea20

Reply:
Ah, push failed because of the added commits. Re-pushed at d90c4b76!